sklearn/neighbors/_base.py

```diff
@@ -619,6 +618,28 @@ def _fit(self, X, y=None):
                 else:
                     self._fit_method = "brute"
 
+        if (
+            self.effective_metric_ in ["wminkowski", "minkowski"]
+            and self.effective_metric_params_["p"] < 1
+        ):
+            # For 0 < p < 1 Minkowski distances aren't valid distance
+            # metric as they do not satisfy triangular inequality:
+            # they are semimetrics.
+            # algo="kd_tree" and algo="ball_tree" can't be used because
+            # KDTree and BallTree require distance metric to work properly.
+            # However, the brute-force algorithm supports semimetrics.
+            if self._fit_method == "brute":
+                warnings.warn(
+                    "Mind that for 0 < p < 1, Minkowski metrics are not distance"
+                    " metric. Continuing the execution with `algo='brute'`."
+                )
+            else:
+                raise ValueError(
+                    'algo="kd_tree" or algo="ball_tree" requires 0 < p < 1 for'
+                    " the Minkowski metric. To resolve this problem either "
+                    "set p >= 1 or algo='brute'"
+                )
```

Comment:
Oops, this mistakes in the message originates from one of my previous comment I think.
```suggestion
            else: # self._fit_method in ("kd_tree", "ball_tree")
                raise ValueError(
                    f'algo="{self._fit_method}" does not support 0 < p < 1 for '
                    "the Minkowski metric. To resolve this problem either "
                    'set p >= 1 or algo="brute".'
                )
```

You will need to update tests accordingly.